### PR TITLE
Removing code to support the Remoting-based CLI

### PR DESCRIPTION
**Pull request by jglick**
As proposed [here](https://groups.google.com/d/msg/jenkinsci-dev/upgi9c5u63o/XL-BJZyLEAAJ). Follows up #2795 by removing the code which that change deprecated. Also ~reverts the JDK-8-specific test module introduced in #3759, as these tests are no longer needed~ leaves the JDK 8 test module but deletes all the ysoserial tests from it.

If accepted, [these docs](https://jenkins.io/doc/book/managing/cli/) would need some updates as well.

@jenkinsci/code-reviewers

### Proposed changelog entries

* The `-remoting` option to the Jenkins CLI has been removed. Commands and command options which required this mode have been removed from Jenkins core and will no longer work in plugins; the `login`, `logout`, `install-tool`, `set-build-parameter`, and `set-build-result` commands have been removed. The `-p` and `--username`/`--password` options are also removed. Some `public` classes and methods which were already deprecated as specific to Remoting mode have been removed, though those likely to be used by plugins have been retained for compatibility.
  * Announcement blog post: https://jenkins.io/blog/2019/01/17/remoting-cli-removed

### CODE REVIEW

**Comment by jtnord** [comment]
@mikecirioli as disturbed can you check our internal CLI for any issues with this?

**Review by jvz — APPROVED** [review]
I love deleting old code.

image

**Review by jeffret-b — APPROVED** [review]
It looks good to me. As I said over on the mailing list I'm in favor of removing this deprecated, problematic capability.

I've looked over the changes and they make sense but I haven't done a full in-depth analysis of each change. The build passes locally and on CI so that's a good indicator.

**Comment by jglick** [comment]
The latest test failure appears to be due to broken UC JSON.

**Review by Wadeck — APPROVED** [review]
:bee: Seems a good idea!

**Review by oleg-nenashev — DISMISSED** [review]
I am -1 w.r.t this change, see the developer mailing list. Other maintainers are welcome to ignore my -1 and merge the pull request taking votes

File: test/pom.xml
Comment (thomasgl-orange) [inline test/pom.xml]: You can remove this dependency on `xalan`, the lib was actually used only for `jenkins.security.security218` tests, which have been removed. This would help later update of the JTH version. See #3872.

Comment (jglick) [reply test/pom.xml]: I would leave that to a separate PR. This is really just a rename of `test-pom/pom.xml` → `test/pom.xml` but Git’s rename detection algorithm does not get it because there was already a `test/pom.xml` in the tree and so it refuses to think of this as a deletion follow by a rename. Unlike with (say) Mercurial the rename is not something explicit and user-controllable in the commit record, just something inferred on demand.

**Comment by batmat** [comment]
For reference, Oleg's email on the ML is https://groups.google.com/forum/#!msg/jenkinsci-dev/upgi9c5u63o/W6UmnDwUDwAJ

Copying my answer here, because it applies to this PR context well too:

-------
I understand your point Oleg, but I feel quite strongly that we shouldn't keep the new test-jdk8 module empty or so for a "just in case" reason.

1) In case we ever need it again, we can very easily then revert the removal using Git, and reintroduce this module when needed.
2) in the meantime, it would be deeply misleading for everyone, and mainly newcomers that there is an empty maven module in the source tree.
-------

**Comment by oleg-nenashev** [comment]
"in the meantime, it would be deeply misleading for everyone, and mainly
newcomers that there is an empty maven module in the source tree." .
Jesse's PR removes not only the Java8-only test module, but the entire
structure which enables creating Java-version-specific tests. Reverting
will be difficult, because Jesse didn't split changes into separate
commits. And there will be likely merge conflicts anyway. This is the
reason why I am -1.

Again, I do not block anybody by my -1. If there is strong consensus
between other contributors, go for it

**Comment by jglick** [comment]
> the entire structure which enables creating Java-version-specific tests

It seemed like a fairly straightforward change in either direction: moving a couple of POMs around and adding a `<profile>`.

> And there will be likely merge conflicts anyway.

Right, I _could_ split off the `test-jdk8` refactoring into a separate commit if that would address your concerns but due to Git behavior as noted in https://github.com/jenkinsci/jenkins/pull/3838#discussion_r252702537 it is unlikely that a future `git revert --no-commit` on that would be any easier than just looking at #3759 and doing something similar, especially since you would anyway want to delete the actual current JDK 8-only tests!

**Comment by batmat** [comment]
IIUC this is great: we have a path forward don't we?

Maybe we could simply:
* Have a new PR, or repurpose this one, that would still do the remoting CLI removal, but only empty the test-jdk8 module from its code instead of removing it
** IIUC again, as Oleg does not disagree on the CLI removal, he would approve that (?)
* Then we possibly file a separate PR (or rather first meet to find an agreement) to do the `test-jdk8` module removal part

WDYT?

**Comment by oleg-nenashev** [comment]
> IIUC again, as Oleg does not disagree on the CLI removal, he would approve that (?)

I would. It just needs to be highlighted in the changelog, because there are still users out there.

> but only empty the test-jdk8 module from its code instead of removing it

I am fine with it either. If @jglick is concerned about the empty module, I am happy to provide another Java 8-only test case

**Comment by jglick** [comment]
> I am happy to provide another Java 8-only test case

You mean, find a retroactive justification?

Anyway, sounds like we have a way of making progress, so I can amend this accordingly, and https://github.com/jenkinsci/jenkins/pull/3872 can adapt.

> It just needs to be highlighted in the changelog

Absolutely! I included proposed text in the PR description, but I would presume that whoever does the weekly changelog editing would want to boldface parts of this or something.

**Review by batmat — APPROVED** [review]

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/CLICommand.java]: Probably we do not care, bu https://github.com/jenkinsci/jenkins.py/blob/cbfc2ec231b447638d2d3e3bf82e3bc7bb160549/python-wrapper/src/main/java/jenkins/python/expoint/CLICommandPW.java

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/CLICommand.java]: :bug: it would break the Jenkins module: https://github.com/jenkinsci/ssh-cli-auth-module/blob/1dc5c9625ed485123216dc995e159c1383eaddef/src/main/java/org/jenkinsci/main/modules/cli/auth/ssh/SshCliAuthenticator.java#L44

**Review by oleg-nenashev — DISMISSED** [review]
It will cause a binary compatibility issue in https://github.com/jenkinsci/ssh-cli-auth-module/blob/1dc5c9625ed485123216dc995e159c1383eaddef/src/main/java/org/jenkinsci/main/modules/cli/auth/ssh/SshCliAuthenticator.java#L44

Also, there are few other binary API breakages in the core which may beed to be investigated

File: test/pom.xml
Comment (jglick) [reply test/pom.xml]: Decided to do this here after all, since removal of ysoserial tests is what permits this dependency to be dropped.

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (jglick) [reply core/src/main/java/hudson/cli/CLICommand.java]: AFAICT `jenkins.py` simply mirrors everything in Jenkins; it is not an actual plugin _using_ the API.

File: core/src/main/java/hudson/cli/CLICommand.java
Comment (jglick) [reply core/src/main/java/hudson/cli/CLICommand.java]: No, because `CliTransportAuthenticator` is no longer used.

Comment (jglick) [reply core/src/main/java/hudson/cli/CLICommand.java]: That brings up a good point, though, which is that this module should be patched to rip out the deprecated stuff, and that can be integrated here. I am preparing a patch for that.

**Review by jeffret-b — APPROVED** [review]
I took a look through it again. It still looks good. This is a valuable cleanup.

**Review by oleg-nenashev — COMMENTED** [review]
LGTM. Assuming it is properly communicated.

@jglick would you be up to write a short blogpost announcing the removal of the functionality and the CLI commands? I doubt we can offer mitigation for that, but it may be possible to offer somebody some hints to revive the functionality as a standalone plugin if they are ready to invest time into it

**Comment by oleg-nenashev** [comment]
@daniel-beck As a Security Officer, are you fine with removing Class deserialization tests?

**Comment by jglick** [comment]
> would you be up to write a short blogpost announcing the removal of the functionality and the CLI commands?

Yes this makes sense, let me think about it.

> hints to revive the functionality as a standalone plugin

I doubt that is feasible, much less desirable, and would not offer any hope of it.

**Comment by batmat** [comment]
`docker run --rm -ti -p 8080:8080 -p 50000:50000 -e ID=3838 jenkins/core-pr-tester` for anyone to test this quite easily.

Though now we have Incrementals, it might even be simpler to use https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/main/jenkins-war/2.164-rc27949.9c80cf12deb0/jenkins-war-2.164-rc27949.9c80cf12deb0.war

**Comment by oleg-nenashev** [comment]
Taking the agreements above, it is no longer blocked
@jglick please feel to proceed with fixing CI and merging

**Comment by jglick** [comment]
The test failures were due to an update center outage, but I will start a new build just to be sure.

**Comment by Wadeck** [comment]
🎉

**Pull request merged**
